// @hyext/hy-ui/dist/components/base.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/**
 * 基础组件
 */
import React from "react";
import { processImageComSrc } from "../common/utils";
import { ImageBackground, ActivityIndicator, Image as RN_Image } from "react-native";
export { View, Text, ScrollView, Easing, Animated } from "react-native";
export function Image(props) {
  var src = props.src,
      mode = props.mode,
      style = props.style,
      rest = _objectWithoutPropertiesLoose(props, ["src", "mode", "style"]);

  var source = processImageComSrc(src);
  var resizeMode = mode;
  return /*#__PURE__*/React.createElement(RN_Image, _extends({
    style: style,
    resizeMode: resizeMode,
    source: source
  }, rest));
}
Image.prefetch = RN_Image.prefetch.bind(RN_Image); // Background Image

export function BackgroundImage(props) {
  var src = props.src,
      rest = _objectWithoutPropertiesLoose(props, ["src"]);

  var source = processImageComSrc(src);
  return /*#__PURE__*/React.createElement(ImageBackground, _extends({
    source: source
  }, rest));
}
export var Indicator = ActivityIndicator; // Indicator
